# Wrangle Tables into Relational Model

Tables:

- Wars
- War_Deaths
- War_LargerConflicts
- War_Locations
- Locations
- Participants
- War_Participants
- Participant_ParticipantTypes
- ParticipantTypes

In [1]:
import pandas as pd

In [2]:
war_par = pd.read_csv("Data/wars_participants.csv")
war = pd.read_csv("Data/wars.csv")
orgs = pd.read_csv("Data/participants_armedorgs.csv")
states = pd.read_csv("Data/participants_states.csv")
locs = pd.read_csv("Data/war_locations.csv")

## Create Wars table

Attributes:

- warID
- warName
- startDate
- endDate
- warDescription
- prevWarID
- nextWarID

In [13]:
war_table = war[['warID', 'eventLabel', 'startDate', 'endDate', 'eventDescription', 
                 'prevWarID', 'nextWarID']].copy().drop_duplicates() \
            .rename(columns={'eventLabel':'warName', 'eventDescription':'warDescription'})

In [15]:
war_table[war_table.duplicated(subset=['warID'], keep=False)]

warID             eventLabel             startDate  \
84    Q877450      Laotian Civil War  1953-11-09T00:00:00Z   
94    Q877450      Laotian Civil War  1959-05-23T00:00:00Z   
243  Q1773984  Batalla del pan dulce  1950-01-01T00:00:00Z   
267  Q1773984  Batalla del pan dulce  1950-10-06T00:00:00Z   
273  Q1773984  Batalla del pan dulce  1950-01-01T00:00:00Z   
308  Q1773984  Batalla del pan dulce  1950-10-06T00:00:00Z   

                  endDate                eventDescription prevWarID nextWarID  
84   1975-12-02T00:00:00Z     1959–1975 civil war in Laos       NaN       NaN  
94   1975-12-02T00:00:00Z     1959–1975 civil war in Laos       NaN       NaN  
243  1951-01-01T00:00:00Z  Batalla entre pacistas y ateos       NaN       NaN  
267  1951-01-01T00:00:00Z  Batalla entre pacistas y ateos       NaN       NaN  
273  1950-10-19T00:00:00Z  Batalla entre pacistas y ateos       NaN       NaN  
308  1950-10-19T00:00:00Z  Batalla entre pacistas y ateos       NaN       NaN

- for Q1773984: drop observations with less specificity (243, 267, 273) in favor of observation with most specificity (308)
- for Q877450: drop 84 based on Wikipedia entry for war, which specifies start date of 23 May 1959

In [16]:
war_table = war_table.drop([84, 243, 267, 273])

In [24]:
war_table.to_csv('Data/Wrangled/war.csv', index=False)

## Create the War tables for multivalued attributes

- War_Deaths
- War_LargerConflicts
- War_Locations

In [26]:
war_deaths = war[['warID', 'casualties']].copy().drop_duplicates().dropna()
war_deaths.to_csv('Data/Wrangled/war_deaths.csv', index=False)

In [49]:
war_locations = war[['warID', 'locationID']].copy().drop_duplicates().dropna()
war_locations.to_csv('Data/Wrangled/war_locations.csv', index=False)

In [30]:
war_largerconflict = war[['warID', 'largerConflictID']].copy().drop_duplicates().dropna()
war_largerconflict.to_csv('Data/Wrangled/war_largerconflicts.csv', index=False)

### Houston, we have a problem

In [43]:
conflictIDs = set(war['largerConflictID'].dropna().unique())
warIDs = set(war['warID'].dropna().unique())
not_wars = conflictIDs - warIDs
len(conflictIDs), len(not_wars)

(75, 41)

Out of 75 larger conflicts, 41 are not wars, and thus will not be present in the wars table, so there cannot be a foreign key to the wars table as is.

In [42]:
pwIDs = set(war['prevWarID'].dropna().unique())
warIDs = set(war['warID'].dropna().unique())
not_wars_pw = pwIDs - warIDs
len(pwIDs), len(not_wars_pw)

(8, 3)

In [46]:
nwIDs = set(war['nextWarID'].dropna().unique())
warIDs = set(war['warID'].dropna().unique())
not_wars_nw = nwIDs - warIDs
len(nwIDs), len(not_wars_nw)

(7, 2)

In [48]:
not_wars_pw, not_wars_nw

({'Q2405009', 'Q2629782', 'Q2992527'}, {'Q2405009', 'Q3446062'})

There's also a problem with some of the "next wars" and "previous wars" - 4 of them aren't in the original war list.

- Q2405009 lacks an "instance of" attribute
- Q2629782 and Q2992527 happen before 1945
- Q3446062 is an instance of conflict, not of war

## Create Locations table

- locationID
- locationName
- locationType

In [64]:
locations = locs[['locationID', 'locationLabel', 'locationDescription']].copy() \
            .rename(columns={'locationLabel':'locationName', 'locationCountryLabel': 'countryName'}) \
            .drop_duplicates()
locations.to_csv('Data/Wrangled/locations.csv', index=False)

In [66]:
locIDl = set(locations['locationID'].dropna().unique())
locIDw = set(war_locations['locationID'].dropna().unique())
locIDmissing = locIDw - locIDl
len(locIDmissing)

0

## Create War_Participants table

- warID
- participantID

In [81]:
war_participants = war_par[['warID', 'participantID']].copy().drop_duplicates().dropna()
war_participants.to_csv('Data/Wrangled/war_participants.csv', index=False)

## Create Participants table

- 

In [68]:
all_par = war_par[['participantID', 'participantLabel']].copy().drop_duplicates().dropna()

In [69]:
all_par

participantID                                   participantLabel
0         Q384535  African Union – United Nations Hybrid Operatio...
1            Q801                                             Israel
2            Q241                                               Cuba
3            Q810                                             Jordan
4           Q1036                                             Uganda
..            ...                                                ...
834          Q813                                         Kyrgyzstan
835       Q935288                                Libyan armed forces
837       Q309474                          Al-Aqsa Martyrs' Brigades
841          Q218                                            Romania
851      Q2000803                     Abdullah Azzam Shaheed Brigade

[495 rows x 2 columns]

In [72]:
orgs_par = orgs[['participantID', 'participantLabel', 'participantDescription', 'inception', 'dissolved']].copy().drop_duplicates()

In [73]:
orgs_par

participantID                         participantLabel  \
0          Q80312         Liberation Tigers of Tamil Eelam   
2          Q58967                    Israel Defense Forces   
3         Q160805  United Nations Interim Force in Lebanon   
4           Q7184                                     NATO   
6          Q11223                  United States Air Force   
..            ...                                      ...   
109      Q1674170                  Turkistan Islamic Party   
111      Q1967703                 National Resistance Army   
112       Q935288                      Libyan armed forces   
113      Q1639447                       Pakistan Air Force   
114      Q1806585                          Lashkar-e-Islam   

                                participantDescription             inception  \
0                   militant organization in Sri Lanka  1976-01-01T00:00:00Z   
2                   combined military forces of Israel  1948-05-26T00:00:00Z   
3                                                  NaN  1978-01-01T00:00:00Z   
4    intergovernmental military alliance of Western...  1949-04-04T00:00:00Z   
6    air warfare branch of the United States Armed ...  1947-01-01T00:00:00Z   
..                                                 ...                   ...   
109                                                NaN  1997-01-01T00:00:00Z   
111  military wing of the National Resistance Movem...  1981-06-01T00:00:00Z   
112                  combined military forces of Libya  1977-01-01T00:00:00Z   
113      air warfare branch of Pakistan's armed forces  1947-08-14T00:00:00Z   
114                                                NaN  2004-01-01T00:00:00Z   

                dissolved  
0    2009-01-01T00:00:00Z  
2                     NaN  
3                     NaN  
4                     NaN  
6                     NaN  
..                    ...  
109                   NaN  
111  1995-01-01T00:00:00Z  
112                   NaN  
113                   NaN  
114                   NaN  

[72 rows x 5 columns]

In [74]:
states_par = states[['participantID', 'participantLabel', 'participantDescription', 'inception', 'dissolved', 'countryCode']].copy().drop_duplicates()

In [75]:
states

participant       participantLabel  \
0     http://www.wikidata.org/entity/Q114                  Kenya   
1     http://www.wikidata.org/entity/Q142                 France   
2      http://www.wikidata.org/entity/Q36                 Poland   
3     http://www.wikidata.org/entity/Q142                 France   
4     http://www.wikidata.org/entity/Q115               Ethiopia   
..                                    ...                    ...   
156   http://www.wikidata.org/entity/Q224                Croatia   
157   http://www.wikidata.org/entity/Q419                   Peru   
158   http://www.wikidata.org/entity/Q971  Republic of the Congo   
159   http://www.wikidata.org/entity/Q774              Guatemala   
160  http://www.wikidata.org/entity/Q1014                Liberia   

                                participantDescription countryCode  \
0                       sovereign state in East Africa         KEN   
1    sovereign state with mainland in Europe and se...         FRA   
2                    sovereign state in Central Europe         POL   
3    sovereign state with mainland in Europe and se...         FRA   
4                               Country in East Africa         ETH   
..                                                 ...         ...   
156                  sovereign state in Central Europe         HRV   
157                   sovereign state in South America         PER   
158                  sovereign state in Central Africa         COG   
159                 sovereign state in Central America         GTM   
160                     sovereign state in West Africa         LBR   

                inception dissolved                              government  \
0    1963-01-01T00:00:00Z       NaN                                     NaN   
1                     NaN       NaN   http://www.wikidata.org/entity/Q49890   
2    1918-11-11T00:00:00Z       NaN  http://www.wikidata.org/entity/Q166747   
3                     NaN       NaN  http://www.wikidata.org/entity/Q188961   
4    1941-05-05T00:00:00Z       NaN  http://www.wikidata.org/entity/Q512187   
..                    ...       ...                                     ...   
156  1991-06-25T00:00:00Z       NaN    http://www.wikidata.org/entity/Q7270   
157  1821-07-28T00:00:00Z       NaN    http://www.wikidata.org/entity/Q7270   
158  1960-01-01T00:00:00Z       NaN                                     NaN   
159  1821-01-01T00:00:00Z       NaN                                     NaN   
160  1822-01-07T00:00:00Z       NaN    http://www.wikidata.org/entity/Q7270   

              governmentLabel  \
0                         NaN   
1    semi-presidential system   
2        parliamentary system   
3            decentralization   
4            federal republic   
..                        ...   
156                  republic   
157                  republic   
158                       NaN   
159                       NaN   
160                  republic   

                                 governmentDescription participantID   govtID  
0                                                  NaN          Q114      NaN  
1                                 system of government          Q142   Q49890  
2                                   form of government           Q36  Q166747  
3    process of redistributing or dispersing functi...          Q142  Q188961  
4    federation of states or territories with a rep...          Q115  Q512187  
..                                                 ...           ...      ...  
156  form of government where people elect leaders ...          Q224    Q7270  
157  form of government where people elect leaders ...          Q419    Q7270  
158                                                NaN          Q971      NaN  
159                                                NaN          Q774      NaN  
160  form of government where people elect leaders ...         Q1014    Q7270  

[161 rows x 11 columns]

In [76]:
all_parIDs = set(all_par['participantID'].dropna().unique())
state_parIDs = set(states_par['participantID'].dropna().unique())
org_parIDs = set(orgs_par['participantID'].dropna().unique())
missing_parIDs = all_parIDs - state_parIDs - org_parIDs

In [78]:
all_par[all_par['participantID'].isin(list(missing_parIDs))]

participantID                                   participantLabel
0         Q384535  African Union – United Nations Hybrid Operatio...
6       Q21476759                      National Forces of Liberation
9         Q130216                 Nagorno-Karabakh Autonomous Oblast
10        Q244165                                            Artsakh
11        Q318657                  Armenian Revolutionary Federation
..            ...                                                ...
819     Q18651204                           Resolute Support Mission
823       Q628461                      Popular Resistance Committees
831       Q818618   Democratic Front for the Liberation of Palestine
837       Q309474                          Al-Aqsa Martyrs' Brigades
851      Q2000803                     Abdullah Azzam Shaheed Brigade

[322 rows x 2 columns]

To-do: determine better categories to fit in all of these other participants

## Create Participant_ParticipantTypes table

- participantID
- participantTypeID

## Create ParticipantTypes table

In [3]:
states_types = states[['govtID','governmentLabel','governmentDescription']].copy().drop_duplicates().dropna(subset=['govtID'])
states_types['category'] = 'state'
states_types = states_types.rename(columns={'govtID':'typeID','governmentLabel':'typeLabel','governmentDescription':'typeDescription'})

In [4]:
orgs_types = orgs[['typeID','typeLabel','typeDescription']].copy().drop_duplicates().dropna(subset=['typeID'])
orgs_types['category'] = 'armed organization'

In [5]:
part_types = pd.concat([states_types, orgs_types])

In [6]:
part_types

typeID                       typeLabel  \
0       Q41614         constitutional monarchy   
1        Q7270                        republic   
2     Q3330103          parliamentary monarchy   
5      Q179164                   unitary state   
6      Q512187                federal republic   
7      Q166747            parliamentary system   
13      Q49892             presidential system   
14      Q49890        semi-presidential system   
15     Q188961                decentralization   
18     Q188759        representative democracy   
25      Q43702                   federal state   
28    Q4198907          parliamentary republic   
36    Q1520223         constitutional republic   
40      Q50686              single-party state   
41    Q5255892           plutocratic oligarchy   
45     Q193852           democratic centralism   
51   Q16711032                       Q16711032   
64     Q184558               absolute monarchy   
75    Q5440547  federal parliamentary republic   
77    Q5433328             family dictatorship   
79     Q181800                           Juche   
85    Q3208952                federal monarchy   
90     Q584683               elective monarchy   
103   Q3043547         parliamentary democracy   
111    Q223832  dominion of the British Empire   
128    Q672729                Islamic republic   
150      Q7269                        monarchy   
0        Q7278                 political party   
1    Q20642011              guerrilla movement   
2      Q772547                    armed forces   
3     Q1456930              peacekeeping force   
4      Q245065  intergovernmental organization   
5     Q1127126               military alliance   
6       Q61883                       air force   
7      Q484652      international organization   
8      Q781132                 military branch   
9    Q17149090              armed organization   
10     Q153936                         militia   
12   Q17127659          terrorist organization   
13    Q1691358                        Q1691358   
17     Q213283              diplomatic mission   
20    Q1673189              irregular military   
23    Q1855859                   popular front   
26   Q15627509           military organization   
33       Q8473                        military   
34     Q207320                    paramilitary   
35     Q234497                  special forces   
37     Q209225             liberation movement   
42    Q7210356          political organisation   
45     Q131401                       caliphate   
46    Q7448178        self-proclaimed monarchy   
48   Q21512251           self-proclaimed state   
52      Q37726                            army   
54     Q176799                   military unit   
73   Q60531252             separatist movement   
75     Q732717          law enforcement agency   
79      Q35535                          police   
81    Q1144991          military reserve force   
86   Q17165317                Joint Task Force   

                                       typeDescription            category  
0    type of monarchy in which power is restricted ...               state  
1    form of government where people elect leaders ...               state  
2                                   form of government               state  
5    state governed as a single unit with a supreme...               state  
6    federation of states or territories with a rep...               state  
7                                   form of government               state  
13                                  form of government               state  
14                                system of government               state  
15   process of redistributing or dispersing functi...               state  
18   democracy where citizens elect a small set of ...               state  
25   state consisting of a union of partially self-...               state  
28   type of republic which operates under a parlia...               state  
36   T